storage-nvmf/src/main/java/org/apache/crail/storage/nvmf/client/NvmfStorageEndpoint.java

```diff
@@ -19,208 +19,229 @@
 	}
 
 	enum Operation {
 		WRITE,
-		READ;
+		READ
 	}
 
-	public StorageFuture Op(Operation op, CrailBuffer buffer, BlockInfo remoteMr, long remoteOffset)
-			throws IOException, InterruptedException {
-		int length = buffer.remaining();
-		if (length > CrailConstants.BLOCK_SIZE){
-			throw new IOException("write size too large " + length);
-		}
-		if (length <= 0){
-			throw new IOException("write size too small, len " + length);
-		}
-		if (buffer.position() < 0){
-			throw new IOException("local offset too small " + buffer.position());
-		}
-		if (remoteOffset < 0){
-			throw new IOException("remote offset too small " + remoteOffset);
-		}
+	void putOperation() {
+		outstandingOperations.decrementAndGet();
+	}
 
-		if (remoteMr.getAddr() + remoteOffset + length > endpoint.getNamespaceSize()){
-			long tmpAddr = remoteMr.getAddr() + remoteOffset + length;
-			throw new IOException("remote fileOffset + remoteOffset + len = " + tmpAddr + " - size = " +
-					endpoint.getNamespaceSize());
+	private boolean tryGetOperation() {
+		int outstandingOperationsOld = outstandingOperations.get();
+		if (outstandingOperationsOld < NvmfStorageConstants.QUEUE_SIZE) {
+			return outstandingOperations.compareAndSet(outstandingOperationsOld, outstandingOperationsOld + 1);
 		}
+		return false;
+	}
 
-//		LOG.info("op = " + op.name() +
-//				", position = " + buffer.position() +
-//				", localOffset = " + buffer.position() +
-//				", remoteOffset = " + remoteOffset +
-//				", remoteAddr = " + remoteMr.getAddr() +
-//				", length = " + length);
-
-		NvmeCommand command = freeCommands.poll();
-		while (command == null) {
-			poll();
-			command = freeCommands.poll();
-		}
+	private static int divCeil(int a, int b) {
+		return (a + b - 1) / b;
+	}
 
-		boolean aligned = NvmfStorageUtils.namespaceSectorOffset(sectorSize, remoteOffset) == 0
-				&& NvmfStorageUtils.namespaceSectorOffset(sectorSize, length) == 0;
-		long lba = NvmfStorageUtils.linearBlockAddress(remoteMr, remoteOffset, sectorSize);
-		StorageFuture future = null;
-		if (aligned) {
-//			LOG.info("aligned");
-			command.setBuffer(buffer.getByteBuffer()).setLinearBlockAddress(lba);
-			switch(op) {
-				case READ:
-					command.read();
-					break;
-				case WRITE:
-					command.write();
-					break;
-			}
-			future = futures[(int)command.getId()] = new NvmfStorageFuture(this, length);
-			command.execute();
-		} else {
-//			LOG.info("unaligned");
-			long alignedLength = NvmfStorageUtils.alignLength(sectorSize, remoteOffset, length);
+	private int getNumLogicalBlocks(CrailBuffer buffer) {
+		return divCeil(buffer.remaining(), getLBADataSize());
+	}
 
-			CrailBuffer stagingBuffer = cache.allocateBuffer();
-			stagingBuffer.limit((int)alignedLength);
+	StorageFuture Op(Operation op, CrailBuffer buffer, BlockInfo blockInfo, long remoteOffset) throws InterruptedException, IOException {
+		assert blockInfo.getAddr() + remoteOffset + buffer.remaining() <= getNamespaceCapacity();
```

Comment:
FYI: Java asserts are disabled by default (https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html#enable-disable) These checks will not catch anything unless the JVM is running with `-ea` flag.

Reply:
Yes, this is meant for debugging purposes only.